tekton-catalog/any-sequencer/cmd/root.go

```diff
@@ -0,0 +1,247 @@
+
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func checkConditions(crs []conditionResult, tr *v1beta1.TaskRun) (string, bool) {
+	for _, cr := range crs {
+		parameters := make(map[string]interface{})
+		for _, result := range cr.results {
+			trLabel := tr.Labels["tekton.dev/pipelineTask"]
+			var found bool
+			for _, taskRunResults := range tr.Status.TaskRunResults {
+				if result == taskRunResults.Name {
+					parameters[`results_`+trLabel+`_`+result] = taskRunResults.Value
+					found = true
+					break
+				}
+			}
+			if !found {
+				log.Printf("The result %s does not exist in taskrun %s.", result, trLabel)
+				return cr.condition, false
+			}
+		}
+		expr, err := govaluate.NewEvaluableExpression(cr.condition)
+		if err != nil {
+			log.Fatal("syntax error:", err)
+		}
+
+		evaluateresult, err := expr.Evaluate(parameters)
+		if err != nil {
+			log.Fatal("evaluate error:", err)
+		}
+
+		if result, ok := evaluateresult.(bool); ok {
+			if result {
+				continue
+			}
+		}
+		return cr.condition, false
+	}
+	return "", true
+}
+
+func watch(cmd *cobra.Command, args []string) {
+	log.Printf("Starting to watch taskrun for '%s' and condition in %s/%s.", taskList, namespace, prName)
+
+	tasks := strings.Split(taskList, ",")
+
+	parse_conditions(conditions, &tasks)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Errorf("Get config of the cluster failed: %+v", err)
+		os.Exit(1)
+	}
+
+	tektonClient, err := tektoncdclientset.NewForConfig(config)
+	if err != nil {
+		log.Errorf("Get client of tekton failed: %+v", err)
+		os.Exit(1)
+	}
+
+	labelSelector := "tekton.dev/pipelineRun=" + strings.TrimSpace(prName)
+
+	var failedTasks []string
+	for {
+		watcher, err := tektonClient.TektonV1beta1().TaskRuns(namespace).Watch(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector})
+
+		if err != nil {
+			log.Printf("TaskRun Watcher error:" + err.Error())
+			log.Printf("Please ensure the service account has permission to get taskRun.")
+			os.Exit(1)
+		}
+
+		for event := range watcher.ResultChan() {
+			taskrun := event.Object.(*v1beta1.TaskRun)
+			taskLabel := taskrun.Labels["tekton.dev/pipelineTask"]
+			if contains(tasks, taskLabel) {
+				if taskrun.Status.GetCondition(apis.ConditionSucceeded).IsTrue() {
+					log.Printf("The TaskRun of %s succeeded.", taskLabel)
+					// check the conditions for the task
+					if conditions, ok := conditionMap[taskLabel]; ok {
+						if condition, ok := checkConditions(conditions, taskrun); !ok {
+							log.Printf("The condition %s for the task %s does not meet.", condition, taskLabel)
+							failedTasks = append(failedTasks, taskLabel)
+						} else {
+							watcher.Stop()
+							os.Exit(0)
+						}
+					} else {
+						watcher.Stop()
+						os.Exit(0)
+					}
+				} else if taskrun.Status.GetCondition(apis.ConditionSucceeded).IsFalse() {
```

Comment:
That's a very deeply-indented block of code, it's a bit difficult to read.

Go idiom is for `if`s to handle fast-exit behaviours, while handling the other ones outside of the `if`. Considering how you call `os.Exit(0)` here, it seems possible to do it the idiomatic way:

```go
taskrunStatus = taskrun.Status.GetCondition(apis.ConditionSucceeded)

if taskrunStatus.IsTrue() {
  conditions, ok := conditionMap[taskLabel]
  if !ok { // no conditions to be passed --> any-sequencer success
    watcher.Stop()
    os.Exit(0)
  }

  condition, ok := checkConditions(conditions, taskrun)
  if ok { // condition passed -->  any-sequencer success
    watcher.Stop()
    os.Exit(0)
  }

  log.Printf("The condition %s for the task %s does not meet.", condition, taskLabel)
}

if !contains(failedTasks, taskLabel) {
  failedTasks = append(failedTasks, taskLabel)
}

if len(failedTasks) >= len(tasks) {
  log.Printf("All specified TaskRun(s) failed.")
  watcher.Stop()
  os.Exit(1)
}
```

Reply:
Thanks for your suggestion, I just modified it based on original code, I agree it's more concise this way, thanks!

Reply:
Tested and we cannot do this, the reason is for the task if the status is running or unknown, it will also be appended to failed task. We should only append to failed task list if the task status is false or the condition failed.